### ISSUE TITLE
"RangeError: Maximum call stack size exceeded" when file includes itself

### ISSUE THREAD

**Issue by AndrewRayCode** [benmosher/eslint-plugin-import#210]
.eslintrc:

``` json
{
    "parserOptions": {
        "sourceType": "module",
    },
    "plugins": [
        "import"
    ],
    "rules": {
        "import/named": 2,
    }
}
```

`import/named` seems to be required for this to fail. Without that rule, eslint exits without error (as in the file passes the check)

Let's create a file that requires itself with the following code inside of it:

``` js
export { test } from './test.js'
```

...and name it test.js:

``` sh
$ eslint --version
    v2.3.0

$ echo "export { test } from './test.js';" > test.js

$ eslint test.js
Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at Array.forEach (native)
    at test/node_modules/eslint-plugin-import/lib/core/getExports.js:243:26
    at test/node_modules/eslint-plugin-import/lib/core/getExports.js:258:13
    at Array.forEach (native)
    at Function.parse (test/node_modules/eslint-plugin-import/lib/core/getExports.js:187:16)
    at Function._for (test/node_modules/eslint-plugin-import/lib/core/getExports.js:139:29)
    at ExportMap.resolveReExport (test/node_modules/eslint-plugin-import/lib/core/getExports.js:74:27)
    at test/node_modules/eslint-plugin-import/lib/core/getExports.js:241:41
    at test/node_modules/eslint-plugin-import/lib/core/getExports.js:258:13
    at Array.forEach (native)
```

Obviously it's a mistake that the file includes itself, but this plugin should warn, not crash, in that case.

**Comment by benmosher** [comment]
Ah, dang. I think this should be fixed when I fix #200, and even if not, I will definitely add as a test case. Thanks!

**Comment by j-f1** [comment]
@benmosher this still happens when you do

```js
// test.js
export * from './test.js'
```

**Comment by ljharb** [comment]
That should be a runtime error; in other words, that code can't ever execute.

However, this eslint plugin shouldn't crash; but any rule touching imports should probably error out on it.

**Comment by j-f1** [comment]
Somehow it’s been working in production for 10 months [in TypeScript in GitHub Desktop](https://github.com/desktop/desktop/blob/d1aa24199b5f14591a88d78f4ebfb55a435659f6/app/src/lib/git/index.ts#L11).

**Comment by ljharb** [comment]
That seems like a Typescript bug then.

**Comment by j-f1** [comment]
Still, it seems like this plugin should parse invalid constructs, since the job of a linter is to warn you before you encounter a runtime (thrown or logic) error.

**Comment by ljharb** [comment]
Totally agree! I'll reopen this; hopefully someone can out in a PR that fixes it.

**Comment by j-f1** [comment]
Looks like this block

https://github.com/benmosher/eslint-plugin-import/blob/68f63f59538ceeb4f0348185683839be94d152c2/src/ExportMap.js#L386-L391

should check

```js
if (remoteMap === m.path) return
```

before calling `ExportMap.for()`.

**Comment by ljharb** [comment]
#449/ #727 may have addressed this; happy to reopen if not.
